src/main/java/impl/image/ConvolutionProviderFactory.java

```diff
@@ -2,8 +2,41 @@
 
 import api.image.ConvolutionProvider;
 
+import java.awt.*;
+import java.lang.Math.*;
 public class ConvolutionProviderFactory {
-    public static ConvolutionProvider getInstance() {
-        return null;
+    private static ConvolutionProviderFactory instance;
+    public static ConvolutionProviderFactory getInstance()
+    {
+        if (instance == null)instance = new ConvolutionProviderFactory();
+        return instance;
     }
+    public Color[][] apply(Color[][] image, double[][] kernel){
+        int n=image.length,m=image[0].length,kn =kernel.length,km=kernel[0].length;
+        Color[][] img = new Color[n][m];
+        for(int i=0;i<n;i++){
+            for(int j=0;j<m;j++){
+                int blue=0,red=0,green=0;
+                int id1=0,id2 =0;
+                for (int it1=-(kn/2);it1<=(kn/2);it1++){
```

Comment:
You use expression `kn/2` , `km/2` expression more than one time, so extract it to a variable